### PR TITLE
sixth assignment

### PR DESCRIPTION
**Pull request by NataliaDymnikova**

### CODE REVIEW

File: assignment6/src/Movement.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    public Movement(int from, int to) {
+        this.from = from;
+        this.to = to;
+    }
+
+    public void print() {
+        logger.info("move from " + from + " to " + to);
+    }
+
+    public Entry<Integer, Integer> getValues() {
```
Comment (simon-frankau) [inline assignment6/src/Movement.java]: I think converting to an Entry is a bad design. A move really isn't a key-value pair, it's swapping of a set of indices. All you're using Entry for is as a way of returning a pair of values - and you could just get the same effect by having accessors "getFrom" and "getTo".

File: assignment6/src/ParkingTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    private void checkEntries(int[][] realMovements, List<Movement> expectedMovements) {
+        int size = expectedMovements.size();
+        assertThat(size, is(realMovements.length));
+        for (int i = 0; i < size; i++) {
+            Entry<Integer, Integer> entry = expectedMovements.get(i).getValues();
+            assertThat(new int[]{entry.getKey(), entry.getValue()}, is(realMovements[i]));
+        }
+    }
```
Comment (simon-frankau) [inline assignment6/src/ParkingTest.java]: A better way to check is to run the moves provided on the input provided and check that it results in the expected target array.

Checking that the moves are precisely the moves you expect is brittle - there are multiple sets of moves that produce the correct result, and if someone else tries to run their algorithm against your tests, the algorithm may be fine and the tests fail. This reduces maintainability - if you tweak the algorithm, the tests may break even if there's no real problem. Try to avoid brittle tests.

File: assignment6/src/ParkingTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+}
```
Comment (simon-frankau) [inline assignment6/src/ParkingTest.java]: You have some good tests on small input, and well-named tests (it's useful to know you're testing the case where zero moves and doesn't move - these are good things to cover), but I would have liked to see some more tests - maybe larger input, or testing a larger array moved cyclically one place left and right, or a long array reversed, or swapped elementwise. Think of things that should stress the algorithm and find corner-case problems, and test them.

File: assignment6/src/Parking.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        }
+        return carsPosition;
+    }
+
+
+    private void swap(int pos1, int pos2, int[] parking) {
+        parking[pos1] ^= parking[pos2];
+        parking[pos2] ^= parking[pos1];
+        parking[pos1] ^= parking[pos2];
+    }
```
Comment (simon-frankau) [inline assignment6/src/Parking.java]: This trick really isn't worth it. Just use a temporary variable. :)

File: assignment6/src/Parking.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    private void swap(int pos1, int pos2, List<Integer> parking) {
```
Comment (simon-frankau) [inline assignment6/src/Parking.java]: Rather than create two swap methods, I'd recommend just putting all your input data into the same format - either all Lists, or all arrays. Maintaining two versions is unnecessary.

File: assignment6/src/Parking.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        for (int i = 0; i < size; i++) {
+            if (beginParking[i] != endParking[i]) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * Create list of cars position from sequence of cars
```
Comment (simon-frankau) [inline assignment6/src/Parking.java:83]: Calling it a "reverse map" or "inverse map" may help people understand what's being produced.

File: assignment6/src/Parking.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+     * <p>
+     * Complexity: O(n)
+     *
+     * @param beginParking -- sequence of cars on the beginning (from zero)
+     * @param endParking   -- sequence of cars on the end
+     * @return -- sequence of movements
+     */
+    public List<Movement> RearrangingCars(int[] beginParking, int[] endParking) {
+        List<Integer> beginCarsPosition = createCarsPositionList(beginParking);
+        List<Movement> movements = new LinkedList<>();
```
Comment (simon-frankau) [inline assignment6/src/Parking.java]: Even for append-only situations, ArrayList does the job fine. LinkedList only really has an advantage if you're trying to add/remove elements from the start/middle of the list.

File: assignment6/src/Parking.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+        while (true) {
+            int posOfZero = beginCarsPosition.get(0);
+            int carToPosOfZero = endParking[posOfZero];
+            int posOfCarToZeroPos = beginCarsPosition.get(carToPosOfZero);
+
+            if (endParking[posOfZero] != 0) {
```
Comment (simon-frankau) [inline assignment6/src/Parking.java]: Use `carToPosOfZero` instead of `endParking[posOfZero]`, since you've already put it in a variable.

File: assignment6/src/Parking.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                makeMovement(posOfZero, posOfCarToZeroPos, beginCarsPosition, beginParking, movements);
+            } else {
+                int posCarToZero = firstPosDifferentCar(beginParking, endParking);
+                if (posCarToZero == -1) {
+                    break;
```
Comment (simon-frankau) [inline assignment6/src/Parking.java]: Using "while(true)" and "break" is a little bit ugly, but given the way the loop's structured, I'll admit I can't see an easy and pleasant way to arrange it differently. Ditto use of a special distinguished value (-1) - ugly, but the easiest way to do it in this context.

File: assignment6/src/Parking.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+     * Return sequence of movement cars (from which places cars move to which).
+     * Function changes beginParking.
+     * <p>
+     * Complexity: O(n)
+     *
+     * @param beginParking -- sequence of cars on the beginning (from zero)
+     * @param endParking   -- sequence of cars on the end
+     * @return -- sequence of movements
+     */
+    public List<Movement> RearrangingCars(int[] beginParking, int[] endParking) {
```
Comment (simon-frankau) [inline assignment6/src/Parking.java:22]: You modify beginParking. You should either copy it or document that you'll modify the input data.

File: assignment6/src/Parking.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /**
+     * Find first car place where car is not on the right place
+     *
+     * @param beginParking -- current sequence of cars
+     * @param endParking   -- sequence of cars on the end
+     * @return -- position of the car, -1 if all cars are on the right places
+     */
+    private int firstPosDifferentCar(int[] beginParking, int[] endParking) {
```
Comment (simon-frankau) [inline assignment6/src/Parking.java]: I think this makes your algorithm potentially quadratic - if you have to call this lots, it's another linear scan inside the outer loop. There are ways to optimise this.

File: assignment6/src/Parking.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        parking[pos2] ^= parking[pos1];
+        parking[pos1] ^= parking[pos2];
+    }
+
+    private void swap(int pos1, int pos2, List<Integer> parking) {
+        parking.set(pos1, parking.get(pos1) ^ parking.get(pos2));
+        parking.set(pos2, parking.get(pos2) ^ parking.get(pos1));
+        parking.set(pos1, parking.get(pos1) ^ parking.get(pos2));
+    }
+}
```
Comment (simon-frankau) [inline assignment6/src/Parking.java]: I see you've done the optimal algorithm, which is a bit trickier than the basic one that some of your colleagues have done. Given that, it's a pretty understandable algorithm. I think there's more polish that could be applied - consistent use of arrays vs. Lists, not modifying input, simplified swap, non-quadratic algorithm, etc., but the core of it is good. Thanks.

**Review by simon-frankau — COMMENTED** [review]
The code could be polished a bit, but the algorithm itself looks good to me. This optimised-for-move-count algorithm is harder work, so thanks for doing it.

File: assignment6/src/Parking.java (hunk 12 above)
Comment (NataliaDymnikova) [reply assignment6/src/Parking.java]: Thank you, Simon!
I am grateful for your comments
